azkaban-common/src/main/java/azkaban/storage/StorageManager.java

```diff
@@ -63,7 +90,23 @@ public void uploadProject(
     );
     log.info(String.format("Adding archive to storage. Meta:%s File: %s[%d bytes]",
         metadata, localFile.getName(), localFile.length()));
-    storage.put(metadata, localFile);
+
+    /* upload to storage */
+    URI uri = storage.put(metadata, localFile);
+
+    /* Add metadata to db */
+    // TODO spyne: remove hack. Database storage should go through the same flow
+    if (!(storage instanceof DatabaseStorage)) {
```

Comment:
No. The way I am approaching this design is that the storage shouldn't be bothered about putting metadata into Azkaban DB. It should simply take care of storing the file and returning a valid uri to reference it. Currently the metadata management is being done inside the `DatabaseStorage`. That's the reason for the hack.